### PR TITLE
E2E UI framework with Cucumber and Thirtyfour

### PR DESCRIPTION
**Pull request by mrrajan**
This pull request combines Thirty Four framework with Cucumber, enabling interaction with web elements and the creation of human-readable test scenarios using Gherkin syntax.

**Thirty Four**
Thirty Four is a Selenium WebDriver library designed for browser automation and UI testing by engaging with web elements, empowering browser-based automation within Rust.

**Gherkin**
Gherkin uses plain-text syntax with Given-When-Then steps to define software behaviors in a human-readable format. For example,
```
  Scenario: Successful Login
    Given I am on the login page
    When I enter valid credentials
    And click the login button
    Then I should be redirected to the dashboard
```
The High level keywords on Gherkin formats written under `/tests/features` directory with `.features` extension and each gherkin steps are defined in low level on Rust under `/src/page` directory with `.rs` extension.

***Note***
This code is tested with localhost upstream version. To test hosted environments (like staging.trustification.dev) - we need a test user with access to Spog UI.

### CODE REVIEW

File: e2e-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# End - End UI Tests
+### Overview
+The End - End UI framework is based on [Cucumber](https://github.com/cucumber-rs/cucumber "Cucumber") and [Thrityfour](https://github.com/stevepryde/thirtyfour/tree/main/thirtyfour "Thrityfour"). The tests are written in Gherkin format and they are executed with the aid of cucumber-rs. 
+
+The tests under package e2e-tests can be run with the below command,
+```shell
+git clone https://github.com/trustification/trustification
+cd e2e-tests 
+cargo run -- --application http://localhost:8084/ --user-name admin --password admin123456
```
Comment (jcrossley3) [inline e2e-tests/README.md:9]: I think you should add a note about how to deploy what's running on `localhost:8084`:

```suggestion
git clone https://github.com/trustification/trustification
# in another shell...
cd trustification/deploy/compose/
podman-compose -f compose.yaml -f compose-guac.yaml -f compose-trustification.yaml up
# ...back to original shell
cd trustification/e2e-tests 
cargo run -- --application http://localhost:8084/ --user-name admin --password admin123456
```

**Review by jcrossley3 — CHANGES_REQUESTED** [review]
I couldn't get it to work locally. My output:
```
[jim@fedora e2e-tests]$ cargo run -- --application http://localhost:8084/ --user-name admin --password admin123456
    Updating crates.io index
    Updating git repository `https://github.com/SeleniumHQ/selenium`
 Downloading crates ...
  Downloaded globset v0.4.14
  Downloaded rustc-demangle v0.1.23
  Downloaded Inflector v0.11.4
  Downloaded vte v0.11.1
  Downloaded toml v0.8.8
  Downloaded serde_spanned v0.6.4
  Downloaded toml_datetime v0.6.5
  Downloaded strip-ansi-escapes v0.2.0
  Downloaded smart-default v0.7.1
  Downloaded is_executable v1.0.1
  Downloaded exitcode v1.1.2
  Downloaded clap_lex v0.6.0
  Downloaded clap_derive v4.4.7
  Downloaded env_logger v0.10.1
  Downloaded arrayvec v0.7.4
  Downloaded terminal_size v0.3.0
  Downloaded backtrace v0.3.69
  Downloaded cfb v0.7.3
  Downloaded addr2line v0.21.0
  Downloaded quick-xml v0.31.0
  Downloaded gimli v0.28.0
  Downloaded clap_builder v4.4.8
  Downloaded inventory v0.3.13
  Downloaded tempfile v3.8.1
  Downloaded ignore v0.4.21
  Downloaded object v0.32.1
  Downloaded lazy-regex-proc_macros v3.1.0
  Downloaded cucumber-codegen v0.20.1
  Downloaded clap v4.4.8
  Downloaded unicode-linebreak v0.1.5
  Downloaded sealed v0.5.0
  Downloaded nom_locate v4.2.0
  Downloaded inflections v1.1.1
  Downloaded junit-report v0.8.3
  Downloaded directories v5.0.1
  Downloaded lazy-regex v3.1.0
  Downloaded infer v0.8.1
  Downloaded synthez-core v0.3.1
  Downloaded filetime_creation v0.1.6
  Downloaded synthez-codegen v0.3.1
  Downloaded drain_filter_polyfill v0.1.3
  Downloaded bytecount v0.6.7
  Downloaded ar v0.9.0
  Downloaded smawk v0.3.2
  Downloaded derive-getters v0.3.0
  Downloaded typed-builder-macro v0.15.2
  Downloaded xz2 v0.1.7
  Downloaded typed-builder v0.15.2
  Downloaded peg v0.6.3
  Downloaded cucumber-expressions v0.3.0
  Downloaded peg-runtime v0.6.3
  Downloaded vte_generate_state_changes v0.1.1
  Downloaded nt-time v0.5.0
  Downloaded infer v0.15.0
  Downloaded peg-macros v0.6.3
  Downloaded lzma-rust v0.1.5
  Downloaded gherkin v0.14.0
  Downloaded synthez v0.3.1
  Downloaded cucumber v0.20.1
  Downloaded toml_edit v0.21.0
  Downloaded debpkg v0.6.0
  Downloaded sevenz-rust v0.5.3
  Downloaded lzma-sys v0.1.20
   Compiling libc v0.2.150
   Compiling serde v1.0.193
   Compiling regex-automata v0.4.3
   Compiling rustix v0.38.25
   Compiling serde_derive v1.0.193
   Compiling tokio-macros v2.2.0
   Compiling linux-raw-sys v0.4.11
   Compiling serde_json v1.0.108
   Compiling crossbeam-utils v0.8.16
   Compiling bstr v1.8.0
   Compiling clap_lex v0.6.0
   Compiling clap_derive v4.4.7
   Compiling crossbeam-epoch v0.9.15
   Compiling crossbeam-deque v0.8.3
   Compiling jobserver v0.1.26
   Compiling num_cpus v1.16.0
   Compiling socket2 v0.5.5
   Compiling mio v0.8.9
   Compiling socket2 v0.4.9
   Compiling filetime v0.2.22
   Compiling xattr v1.0.1
   Compiling parking_lot_core v0.9.8
   Compiling dirs-sys v0.4.1
   Compiling console v0.15.7
   Compiling cc v1.0.83
   Compiling tar v0.4.40
   Compiling filetime_creation v0.1.6
   Compiling sevenz-rust v0.5.3
   Compiling tokio v1.34.0
   Compiling parking_lot v0.12.1
   Compiling directories v5.0.1
   Compiling ring v0.16.20
   Compiling zstd-sys v2.0.8+zstd.1.5.5
   Compiling bzip2-sys v0.1.11+1.0.8
   Compiling backtrace v0.3.69
   Compiling lzma-sys v0.1.20
   Compiling regex v1.10.2
   Compiling globset v0.4.14
   Compiling terminal_size v0.3.0
   Compiling clap_builder v4.4.8
   Compiling xz2 v0.1.7
   Compiling ignore v0.4.21
   Compiling cucumber-expressions v0.3.0
   Compiling anyhow v1.0.75
   Compiling is-terminal v0.4.9
   Compiling env_logger v0.10.1
   Compiling lazy-regex-proc_macros v3.1.0
   Compiling stringmatch v0.4.0
   Compiling globwalk v0.8.1
   Compiling tempfile v3.8.1
   Compiling cucumber-codegen v0.20.1
   Compiling bzip2 v0.4.4
   Compiling rustls v0.21.7
   Compiling clap v4.4.8
   Compiling lazy-regex v3.1.0
   Compiling tokio-util v0.7.9
   Compiling h2 v0.3.21
   Compiling toml_datetime v0.6.5
   Compiling serde_spanned v0.6.4
   Compiling serde_urlencoded v0.7.1
   Compiling rustls-webpki v0.101.6
   Compiling sct v0.7.0
   Compiling toml_edit v0.21.0
   Compiling webdriver v0.48.0
   Compiling gherkin v0.14.0
   Compiling toml v0.8.8
   Compiling zstd-safe v5.0.2+zstd.1.5.2
   Compiling cucumber v0.20.1
   Compiling zstd v0.11.2+zstd.1.5.2
   Compiling hyper v0.14.27
   Compiling debpkg v0.6.0
   Compiling zip v0.6.6
   Compiling tokio-rustls v0.24.1
   Compiling hyper-rustls v0.24.1
   Compiling fantoccini v0.20.0-rc.7
   Compiling reqwest v0.11.22
   Compiling thirtyfour v0.32.0-rc.8
   Compiling selenium-manager v0.4.16 (https://github.com/SeleniumHQ/selenium?branch=trunk#dde3eed5)
   Compiling e2e-tests v0.1.0 (/home/jim/src/trustification/trustification/e2e-tests)
warning: associated function `new` is never used
  --> e2e-tests/src/world.rs:38:14
   |
37 | impl E2EWorld {
   | ------------- associated function in this implementation
38 |     async fn new() -> io::Result<Self> {
   |              ^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: `e2e-tests` (bin "e2e-tests") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 22.46s
     Running `/home/jim/src/trustification/trustification/target/debug/e2e-tests --application 'http://localhost:8084/' --user-name admin --password admin123456`
Configuring Webdriver, It may take sometime. Please wait...
Driver Path from script ./driver/chromedriver/linux64/119.0.6045.105/chromedriver
driver path ./driver/chromedriver/linux64/119.0.6045.105/chromedriver
Loading Test data for localhost, it might take sometime...
Starting ChromeDriver 119.0.6045.105 (38c72552c5e15ba9b3117c0967a0fd105072d7c6-refs/branch-heads/6045@{#1103}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
Command failed with an error: Ok(" Downloading crates ...\n  Downloaded clap_derive v3.2.25\n  Downloaded semver v1.0.19\n  Downloaded names v0.14.0\n   Compiling serde v1.0.193\n   Compiling getrandom v0.2.10\n   Compiling signal-hook-registry v1.4.1\n   Compiling mio v0.8.9\n   Compiling serde_derive v1.0.193\n   Compiling tokio-macros v2.2.0\n   Compiling serde_json v1.0.108\n   Compiling openssl-sys v0.9.93\n   Compiling ring v0.16.20\n   Compiling zstd-sys v2.0.8+zstd.1.5.5\n   Compiling utoipa-gen v4.0.0\n   Compiling clap_builder v4.4.8\n   Compiling clap_derive v4.4.7\n   Compiling psm v0.1.21\n   Compiling ahash v0.8.3\n   Compiling ahash v0.7.6\n   Compiling rand_core v0.6.4\n   Compiling hashbrown v0.12.3\n   Compiling hashbrown v0.14.1\n   Compiling rand_chacha v0.3.1\n   Compiling stacker v0.1.15\n   Compiling libz-sys v1.1.12\n   Compiling validator_derive v0.16.0\n   Compiling rand v0.8.5\n   Compiling dirs-sys v0.3.7\n   Compiling indexmap v1.9.3\n   Compiling openssl v0.10.57\n   Compiling native-tls v0.2.11\n   Compiling tokio-openssl v0.6.3\n   Compiling dirs v4.0.0\n   Compiling sct v0.7.0\n   Compiling rustls-webpki v0.101.6\n   Compiling indexmap v2.0.2\n   Compiling webpki v0.22.2\n   Compiling prometheus v0.13.3\n   Compiling tokio v1.34.0\n   Compiling chumsky v1.0.0-alpha.6\n   Compiling shellexpand v2.1.2\n   Compiling opentelemetry_api v0.20.0\n   Compiling rust-embed-impl v8.0.0\n   Compiling rustls v0.21.7\n   Compiling utoipa-swagger-ui v4.0.0\n   Compiling libssh2-sys v0.2.23\n   Compiling tracing-subscriber v0.3.17\n   Compiling threadpool v1.8.1\n   Compiling libgit2-sys v0.14.2+1.5.1\n   Compiling rust-embed v8.0.0\n   Compiling thrift v0.17.0\n   Compiling gethostname v0.2.3\n   Compiling clap v4.4.8\n   Compiling petgraph v0.6.4\n   Compiling env_logger v0.10.1\n   Compiling strum_macros v0.24.3\n   Compiling rustls v0.20.9\n   Compiling cpe v0.1.3 (https://github.com/trustification/cpe-rs?rev=dc2c5661d436410cc9562596ab16a2e771261eb2#dc2c5661)\n   Compiling ordered-multimap v0.4.3\n   Compiling dirs v5.0.1\n   Compiling rust-ini v0.18.0\n   Compiling lru v0.11.1\n   Compiling fs4 v0.6.6\n   Compiling memmap2 v0.7.1\n   Compiling toml_edit v0.19.15\n   Compiling io-lifetimes v1.0.11\n   Compiling polling v2.8.0\n   Compiling rustix v0.37.24\n   Compiling cmake v0.1.50\n   Compiling dirs-sys-next v0.1.2\n   Compiling dirs-next v2.0.0\n   Compiling signatory v0.23.2\n   Compiling term v0.7.0\n   Compiling rdkafka-sys v4.6.0+2.2.0\n   Compiling nkeys v0.2.0\n   Compiling nuid v0.3.2\n   Compiling futures-intrusive v0.5.0\n   Compiling libsqlite3-sys v0.26.0\n   Compiling proc-macro-crate v1.3.1\n   Compiling tokio-util v0.7.9\n   Compiling tokio-stream v0.1.14\n   Compiling tokio-native-tls v0.3.1\n   Compiling tokio-rustls v0.24.1\n   Compiling actix-rt v2.9.0\n   Compiling tokio-rustls v0.23.4\n   Compiling h2 v0.3.21\n   Compiling actix-codec v0.5.1\n   Compiling actix-tls v3.1.1\n   Compiling actix-server v2.3.0\n   Compiling tower v0.4.13\n   Compiling opentelemetry_sdk v0.20.0\n   Compiling aws-smithy-async v0.55.3\n   Compiling async-io v1.13.0\n   Compiling tokio-io-timeout v1.2.0\n   Compiling tokio-retry v0.3.0\n   Compiling async-nats-tokio-rustls-deps v0.24.0-ALPHA.1\n   Compiling async-global-executor v2.3.1\n   Compiling num_enum_derive v0.5.11\n   Compiling string_cache v0.8.7\n   Compiling async-std v1.12.0\n   Compiling deranged v0.3.9\n   Compiling url v2.4.1\n   Compiling serde_urlencoded v0.7.1\n   Compiling chrono v0.4.31\n   Compiling either v1.9.0\n   Compiling actix-router v0.5.1\n   Compiling serde_yaml v0.9.25\n   Compiling hex v0.4.3\n   Compiling time v0.3.30\n   Compiling uuid v1.4.1\n   Compiling packageurl v0.3.0\n   Compiling itertools v0.10.5\n   Compiling opentelemetry v0.20.0\n   Compiling humantime-serde v1.1.1\n   Compiling validator v0.16.1\n   Compiling biscuit v0.6.0\n   Compiling schemars v0.8.15\n   Compiling actix-web-codegen v4.2.2\n   Compiling opentelemetry-semantic-conventions v0.12.0\n   Compiling utoipa v4.0.0\n   Compiling toml_datetime v0.6.5\n   Compiling serde_spanned v0.6.4\n   Compiling opentelemetry-jaeger v0.19.0\n   Compiling cookie v0.16.2\n   Compiling sikula v0.4.2\n   Compiling hyper v0.14.27\n   Compiling tracing-bunyan-formatter v0.3.9\n   Compiling tracing-opentelemetry v0.20.0\n   Compiling aws-smithy-types v0.55.3\n   Compiling rayon v1.8.0\n   Compiling cvss v2.0.0\n   Compiling semver v1.0.19\n   Compiling smol_str v0.2.0\n   Compiling spdx-expression v0.5.2\n   Compiling cyclonedx-bom v0.4.0 (https://github.com/trustification/cyclonedx-rust-cargo?branch=trustification#63d912ad)\n   Compiling itertools v0.11.0\n   Compiling platforms v3.1.2\n   Compiling spdx-rs v0.5.5\n   Compiling bytes-utils v0.1.3\n   Compiling serde_with v3.3.0\n   Compiling toml v0.7.8\n   Compiling prost-derive v0.11.9\n   Compiling tantivy-common v0.6.0\n   Compiling attohttpc v0.24.1\n   Compiling cargo-lock v9.0.0\n   Compiling quick-xml v0.28.2\n   Compiling aws-credential-types v0.55.3\n   Compiling tantivy-stacker v0.2.0\n   Compiling tantivy-tokenizer-api v0.2.0\n   Compiling rust-stemmers v1.2.0\n   Compiling sketches-ddsketch v0.2.1\n   Compiling v11y-model v0.1.0 (/home/jim/src/trustification/trustification/v11y/model)\n   Compiling combine v3.8.1\n   Compiling which v4.4.2\n   Compiling aws-smithy-json v0.55.3\n   Compiling graphql-introspection-query v0.2.0\n   Compiling hyper-tls v0.5.0\n   Compiling hyper-rustls v0.24.1\n   Compiling aws-smithy-http v0.55.3\n   Compiling reqwest v0.11.22\n   Compiling bombastic-model v0.1.0 (/home/jim/src/trustification/trustification/bombastic/model)\n   Compiling hyper-rustls v0.23.2\n   Compiling aws-creds v0.35.0\n   Compiling prost v0.11.9\n   Compiling aws-smithy-query v0.55.3\n   Compiling axum v0.6.20\n   Compiling hyper-timeout v0.4.1\n   Compiling num_enum v0.5.11\n   Compiling rust-s3 v0.34.0-beta3 (https://github.com/trustification/rust-s3.git?branch=trustification#81b6f172)\n   Compiling serde_nanos v0.1.3\n   Compiling openvex v0.1.1\n   Compiling async-nats v0.29.0\n   Compiling aws-smithy-http-tower v0.55.3\n   Compiling prost-types v0.11.9\n   Compiling aws-sigv4 v0.55.3\n   Compiling aws-smithy-client v0.55.3\n   Compiling openid v0.12.1\n   Compiling trustification-common v0.1.0 (/home/jim/src/trustification/trustification/common)\n   Compiling trustification-api v0.1.0 (/home/jim/src/trustification/trustification/api)\n   Compiling aws-types v0.55.3\n   Compiling prost-build v0.11.9\n   Compiling aws-sig-auth v0.55.3\n   Compiling aws-http v0.55.3\n   Compiling aws-endpoint v0.55.3\n   Compiling ascii-canvas v3.0.0\n   Compiling sqlformat v0.2.2\n   Compiling graphql-parser v0.4.0\n   Compiling tonic-build v0.9.2\n   Compiling aws-sdk-sso v0.28.0\n   Compiling aws-sdk-sts v0.28.0\n   Compiling aws-sdk-sqs v0.27.0\n   Compiling lalrpop v0.19.12\n   Compiling hashlink v0.8.4\n   Compiling console v0.15.7\n   Compiling guac v0.1.0 (https://github.com/trustification/guac-rs.git?rev=7596de7cfafb59255bb6f629fdc9857c4e620fb9#7596de7c)\n   Compiling graphql_client_codegen v0.12.0\n   Compiling tonic v0.9.2\n   Compiling aws-config v0.55.3\n   Compiling vexination-model v0.1.0 (/home/jim/src/trustification/trustification/vexination/model)\n   Compiling sqlx-core v0.7.2\n   Compiling buffered-reader v1.2.0\n   Compiling indicatif v0.17.7\n   Compiling csv v1.3.0\n   Compiling flexible-time v0.1.1\n   Compiling cve v0.2.0 (https://github.com/ctron/cve?rev=77341768d893098f1f8b1397364fecfb16b4d776#77341768)\n   Compiling exhort-model v0.1.0 (/home/jim/src/trustification/trustification/exhort/model)\n   Compiling sha256 v1.4.0\n   Compiling clap_derive v3.2.25\n   Compiling atty v0.2.14\n   Compiling serde_with v2.3.3\n   Compiling segment v0.2.3\n   Compiling vergen v8.2.5\n   Compiling trustification-version v0.1.0 (/home/jim/src/trustification/trustification/version)\n   Compiling actix v0.13.1\n   Compiling trustification-analytics v0.1.0 (/home/jim/src/trustification/trustification/analytics)\n   Compiling spog-api v0.1.0 (/home/jim/src/trustification/trustification/spog/api)\n   Compiling okapi v0.7.0-rc.1\n   Compiling sqlx-sqlite v0.7.2\n   Compiling snyk v0.1.0 (https://github.com/dejanb/snyk-rs.git?branch=main#c889b573)\n   Compiling clap v3.2.25\n   Compiling sqlx-macros-core v0.7.2\n   Compiling graphql_query_derive v0.12.0\n   Compiling graphql_client v0.12.0\n   Compiling sqlx-macros v0.7.2\n   Compiling names v0.14.0\n   Compiling sequoia-openpgp v1.16.1\n   Compiling sqlx v0.7.2\n   Compiling zstd-safe v6.0.6\n   Compiling zstd v0.12.4\n   Compiling actix-http v3.4.0\n   Compiling async-compression v0.4.3\n   Compiling tantivy-sstable v0.2.0\n   Compiling tantivy-columnar v0.2.0\n   Compiling rdkafka v0.31.0\n   Compiling tantivy v0.21.0\n   Compiling trustification-event-bus v0.1.0 (/home/jim/src/trustification/trustification/event-bus)\n   Compiling actix-web v4.4.0\n   Compiling actix-web-extras v0.1.0\n   Compiling actix-web-httpauth v0.8.1\n   Compiling actix-web-prom v0.7.0\n   Compiling actix-web-opentelemetry v0.15.0\n   Compiling actix-cors v0.6.4\n   Compiling actix-ws v0.2.5\n   Compiling git2 v0.16.1\n   Compiling trustification-auth v0.1.0 (/home/jim/src/trustification/trustification/auth)\n   Compiling trustification-infrastructure v0.1.0 (/home/jim/src/trustification/trustification/infrastructure)\n   Compiling collectorist-client v0.1.0 (/home/jim/src/trustification/trustification/collectorist/client)\n   Compiling collector-client v0.1.0 (/home/jim/src/trustification/trustification/collector/client)\n   Compiling crates-index v0.19.13\n   Compiling walker-common v0.5.3\n   Compiling rustsec v0.27.0\n   Compiling v11y-client v0.1.0 (/home/jim/src/trustification/trustification/v11y/client)\n   Compiling collectorist-api v0.1.0 (/home/jim/src/trustification/trustification/collectorist/api)\n   Compiling trustification-admin v0.1.0 (/home/jim/src/trustification/trustification/admin)\n   Compiling csaf-walker v0.5.3\n   Compiling csaf v0.5.0 (https://github.com/voteblake/csaf-rs?rev=76cb9ede10adb1fbb495b17e5fd8d95c5cf6c900#76cb9ede)\n   Compiling sbom-walker v0.5.3\n   Compiling exhort-api v0.1.0 (/home/jim/src/trustification/trustification/exhort/api)\n   Compiling collector-osv v0.1.0 (/home/jim/src/trustification/trustification/collector/osv)\n   Compiling collector-snyk v0.1.0 (/home/jim/src/trustification/trustification/collector/snyk)\n   Compiling collectorist v0.1.0 (/home/jim/src/trustification/trustification/collectorist/collectorist)\n   Compiling trustification-storage v0.1.0 (/home/jim/src/trustification/trustification/storage)\n   Compiling spog-model v0.1.0 (/home/jim/src/trustification/trustification/spog/model)\n   Compiling vexination-walker v0.1.0 (/home/jim/src/trustification/trustification/vexination/walker)\n   Compiling bombastic-walker v0.1.0 (/home/jim/src/trustification/trustification/bombastic/walker)\n   Compiling exhort v0.1.0 (/home/jim/src/trustification/trustification/exhort/exhort)\n   Compiling collector v0.1.0 (/home/jim/src/trustification/trustification/collector/collector)\n   Compiling trustification-index v0.1.0 (/home/jim/src/trustification/trustification/index)\n   Compiling exporter v0.1.0 (/home/jim/src/trustification/trustification/exporter)\n   Compiling trustification-indexer v0.1.0 (/home/jim/src/trustification/trustification/indexer)\n   Compiling v11y-index v0.1.0 (/home/jim/src/trustification/trustification/v11y/index)\n   Compiling vexination-index v0.1.0 (/home/jim/src/trustification/trustification/vexination/index)\n   Compiling bombastic-index v0.1.0 (/home/jim/src/trustification/trustification/bombastic/index)\n   Compiling v11y-indexer v0.1.0 (/home/jim/src/trustification/trustification/v11y/indexer)\n   Compiling v11y-api v0.1.0 (/home/jim/src/trustification/trustification/v11y/api)\n   Compiling bombastic-api v0.1.0 (/home/jim/src/trustification/trustification/bombastic/api)\n   Compiling bombastic-indexer v0.1.0 (/home/jim/src/trustification/trustification/bombastic/indexer)\n   Compiling vexination-indexer v0.1.0 (/home/jim/src/trustification/trustification/vexination/indexer)\n   Compiling vexination-api v0.1.0 (/home/jim/src/trustification/trustification/vexination/api)\n   Compiling v11y-walker v0.1.0 (/home/jim/src/trustification/trustification/v11y/walker)\n   Compiling spog v0.1.0 (/home/jim/src/trustification/trustification/spog/spog)\n   Compiling bombastic v0.1.0 (/home/jim/src/trustification/trustification/bombastic/bombastic)\n   Compiling vexination v0.1.0 (/home/jim/src/trustification/trustification/vexination/vexination)\n   Compiling v11y v0.1.0 (/home/jim/src/trustification/trustification/v11y/v11y)\n   Compiling trust v0.1.0 (/home/jim/src/trustification/trustification/trust)\n    Finished dev [unoptimized + debuginfo] target(s) in 1m 39s\n     Running `/home/jim/src/trustification/trustification/target/debug/trust vexination walker --devmode -3 --sink 'http://localhost:8081/api/v1/vex' --source ../data/ds1/csaf`\n[2023-11-30T14:16:52.933Z ERROR trustification_auth::client::inject] error=InvalidClient: Invalid client or Invalid client credentials\nthread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OpenId(ClientError(OAuth2(OAuth2Error { error: InvalidClient, error_description: Some(\"Invalid client or Invalid client credentials\"), error_uri: None })))', /home/jim/src/trustification/trustification/vexination/walker/src/server.rs:57:30\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n")
[jim@fedora e2e-tests]$ 
```

File: e2e-tests/README.md (hunk 1 above)
Comment (mrrajan) [reply e2e-tests/README.md:10]: @jcrossley3 updated! Please let me know :)

**Comment by jcrossley3** [comment]
Still doesn't work for me...

```
[jim@fedora e2e-tests]$ cargo run -- --application http://localhost:8084/ --user-name admin --password admin123456
warning: associated function `new` is never used
  --> e2e-tests/src/world.rs:38:14
   |
37 | impl E2EWorld {
   | ------------- associated function in this implementation
38 |     async fn new() -> io::Result<Self> {
   |              ^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: `e2e-tests` (bin "e2e-tests") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `/home/jim/src/trustification/trustification/target/debug/e2e-tests --application 'http://localhost:8084/' --user-name admin --password admin123456`
Configuring Webdriver, It may take sometime. Please wait...
Driver Path from script ./driver/chromedriver/linux64/119.0.6045.105/chromedriver
driver path ./driver/chromedriver/linux64/119.0.6045.105/chromedriver
Loading Test data for localhost, it might take sometime...
Starting ChromeDriver 119.0.6045.105 (38c72552c5e15ba9b3117c0967a0fd105072d7c6-refs/branch-heads/6045@{#1103}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
Command failed with an error: Ok("    Finished dev [unoptimized + debuginfo] target(s) in 0.25s\n     Running `/home/jim/src/trustification/trustification/target/debug/trust vexination walker --devmode -3 --sink 'http://localhost:8081/api/v1/vex' --source ../data/ds1/csaf`\n[2023-12-01T14:32:04.245Z ERROR trustification_auth::client::inject] error=InvalidClient: Invalid client or Invalid client credentials\nthread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OpenId(ClientError(OAuth2(OAuth2Error { error: InvalidClient, error_description: Some(\"Invalid client or Invalid client credentials\"), error_uri: None })))', /home/jim/src/trustification/trustification/vexination/walker/src/server.rs:57:30\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n")
[jim@fedora e2e-tests]$ cargo run -- --application http://localhost:8084/ --user-name admin --password admin123456
warning: associated function `new` is never used
  --> e2e-tests/src/world.rs:38:14
   |
37 | impl E2EWorld {
   | ------------- associated function in this implementation
38 |     async fn new() -> io::Result<Self> {
   |              ^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: `e2e-tests` (bin "e2e-tests") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `/home/jim/src/trustification/trustification/target/debug/e2e-tests --application 'http://localhost:8084/' --user-name admin --password admin123456`
Configuring Webdriver, It may take sometime. Please wait...
Driver Path from script ./driver/chromedriver/linux64/119.0.6045.105/chromedriver
driver path ./driver/chromedriver/linux64/119.0.6045.105/chromedriver
Loading Test data for localhost, it might take sometime...
Starting ChromeDriver 119.0.6045.105 (38c72552c5e15ba9b3117c0967a0fd105072d7c6-refs/branch-heads/6045@{#1103}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
[1701441143.213][SEVERE]: bind() failed: Address already in use (98)
IPv6 port not available. Exiting...
Command failed with an error: Ok("    Finished dev [unoptimized + debuginfo] target(s) in 0.25s\n     Running `/home/jim/src/trustification/trustification/target/debug/trust vexination walker --devmode -3 --sink 'http://localhost:8081/api/v1/vex' --source ../data/ds1/csaf`\n[2023-12-01T14:32:23.542Z ERROR trustification_auth::client::inject] error=InvalidClient: Invalid client or Invalid client credentials\nthread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OpenId(ClientError(OAuth2(OAuth2Error { error: InvalidClient, error_description: Some(\"Invalid client or Invalid client credentials\"), error_uri: None })))', /home/jim/src/trustification/trustification/vexination/walker/src/server.rs:57:30\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n")
[jim@fedora e2e-tests]$ 
```

File: e2e-tests/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
+# End - End UI Tests
+### Overview
+The End - End UI framework is based on [Cucumber](https://github.com/cucumber-rs/cucumber "Cucumber") and [Thrityfour](https://github.com/stevepryde/thirtyfour/tree/main/thirtyfour "Thrityfour"). The tests are written in Gherkin format and they are executed with the aid of cucumber-rs. 
+
+For example, the tests under package e2e-tests can be run with the below command to test localhost. 
+(Make sure the localhost is up and running with the [podman](#Setting up Localhost))
+```shell
+git clone https://github.com/trustification/trustification
+cd e2e-tests
```
Comment (jcrossley3) [inline e2e-tests/README.md:9]: I'm still bothered by the implication that running the tests is easy as cloning the repo. Maybe just assume whoever is running them has already cloned the repo?
```suggestion
From the directory containing this README file, run the following:
```

File: e2e-tests/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+So with this we can write our tests and its validation steps under features directory and the definitions for the test steps implemented under rust files with the aid of thirtyfour crate. 
+
+#### Setting up Localhost
+
+Make sure the localhost is up and running to execute tests on ***http://localhost:8084/***. To setup localhost run the below commands on another terminal,
+```shell
+# On another terminal
+git clone https://github.com/trustification/trustification
+cd deploy/compose
```
Comment (jcrossley3) [inline e2e-tests/README.md:110]:
```suggestion
cd ../deploy/compose
```

**Review by jcrossley3 — COMMENTED** [review]
There must be something I'm missing. I follow your steps, but the tests don't run.

**Review by jcrossley3 — COMMENTED** [review]
I'm not ignoring you, @mrrajan -- I genuinely don't know whether it's a good idea for us to maintain selenium tests in the repo. I also think the structure of our basic functionality will change, perhaps significantly.

Whatever we do -- whether these tests live in here or in their own repo -- I believe we need to figure out how to make CI run them. Do you think that's possible in this PR? Can you add a github action to this?

**Comment by mrrajan** [comment]
Closing this PR as we decide to have E2E UI scripts on its own repo.

**Pull request closed without merging**